docs/reference/c++_support.rst

```diff
@@ -0,0 +1,165 @@
+.. meta::
+  :description: This chapter describes the C++ support of the HIP ecosystem
+  ROCm software.
+  :keywords: AMD, ROCm, HIP, C++
+
+*******************************************************************************
+C++ language support
+*******************************************************************************
+
+.. _language_introduction:
+Introduction
+===============================================================================
+
+The main way to harness the power of the ROCm platform is through the usage C++ and HIP
+code. This code is then compiled with a ``clang`` or ``clang++`` compiler. The official
+versions support the HIP platform, but for the most up-to-date feature set use the
+``amdclang`` or ``amdclang++`` installed with your ROCm installation.
+
+The source code will be processed based on the ``C++03``, ``C++11``, ``C++14``, ``C++17``
+or ``C++20`` standards, but supports HIP specific extensions and is subject to certain
+restrictions. The largest restriction is the lack of support of the standard library.
+This is mostly because the SIMD nature of the HIP device makes most of the standard
+library implementations not performant or useful. The important operations are
+implemented in HIP specific libraries like rocPRIM, rocThrust and hipCUB.
+
+.. _language_modern_c++_support:
+Modern C++ support
+===============================================================================
+
+The C++11 and later standards had huge additions to C++ making it a truly modern
+language. In this section, we'll describe which of these features HIP supports.
+
+C++11 support
+-------------------------------------------------------------------------------
+
+The C++11 standard introduced a myriad of new features to the language. These features
+are supported in HIP device code, with some notable omissions. The biggest of which is
+the lack of concurrency support on the device. This is because the HIP device concurrency
+model is fundamentally different compared to the C++ model, which is used on the host
+side. E.G: it doesn't make sense to start a new thread on the device.
+
+There are some restrictions and clarifications for certain features. For example, some
+new features were like functions using ``initializer lists``, ``std::move`` and
+``std::forward`` or functions with ``constexpr`` qualifier are implicitly considered to
+have ``__host__`` and ``__device__`` execution space specifier. Also, ``constexpr``
+variables that are static member variables or namespace scoped can be used from both host
+and device, but only for value access. Dereferencing a static ``constexpr`` on a different
+execution space than specified will cause an error.
```

Comment:
Certain features have restrictions and clarifications. For example, any functions using the ``constexpr`` qualifier or the new ``initializer lists``, ``std::move`` or ``std::forward`` features are implicitly considered to have the ``__host__`` and ``__device__`` execution space specifier.  Also, ``constexpr`` variables that are static members or namespace scoped can be used from both host and device, but only for read access. Dereferencing a static ``constexpr`` outside its specified execution space causes an error.